### ISSUE TITLE
[feature] allow jobs to fail

### ISSUE THREAD

**Issue by yarnabrina** [kubeflow/pipelines#7330]
### Feature Area

/area backend

### What feature would you like to see?

_**Allow failure of jobs**_ - If an operation fails, do not fail the pipeline. Allow the pipeline to continue to the next stage, and there it may fail if that does not have the pre-requisites.

### What is the use case or pain point?

In the machine learning pipelines, it is fairly common to run multiple models, or possibly different configurations of same model, and this possibly runs on a subset of training data. After these are trained, usually they are compared using some metric, the best model is chosen, and that is run on the entire training data to have the final trained model.

If someone uses [`kfp.dsl.ParallelFor`](https://kubeflow-pipelines.readthedocs.io/en/latest/source/kfp.dsl.html#kfp.dsl.ParallelFor) to run the different models, failure in one of them causes the entire pipeline to fail and possibly successful training of other models are lost. But if the next stage, the one to compare using metric supports comparison of the available (i.e. successful) models, the pipeline failure costs the time to train those models, as one have to restart. If we support the requested feature, the failed operations will display an warning (may be ⚠️), and will go on to final training step. Then depending on whether that supports comparison of subset of all models, it will proceed as if the failed models were not there. If not, it'll fail there.

Very similar functionality in available in few CI tools. For example, Gitlab CI has [`allow_failure`](https://docs.gitlab.com/ee/ci/yaml/#allow_failure), Travis CI has [`allow_failures`](https://docs.travis-ci.com/user/customizing-the-build/#jobs-that-are-allowed-to-fail), etc.

### Is there a workaround currently?

It is possible to do very broad top level exception handling to suppress failures. However, in this way the fact that it failed is hidden in the logs and not displayed in the pipeline dashboard. In scheduled pipelines where no one really go through the logs of all "successful" pipelines, these failures will go unnoticed.

---

Love this idea? Give it a 👍. We prioritize fulfilling features with the most 👍.

**Comment by chensun** [comment]
Making it configurable sounds fine to me. 

A possible workaround for now, albeit hacky, you could put your training into an exit handler task. In that case, it would run regardless whether the upstream tasks succeed or not.

**Comment by yarnabrina** [comment]
Hi @chensun, thanks for taking this. Would be very nice to have this.

Regarding the workaround you suggested, can you please give some more details? I did find about `kfp.dsl.ExitHandler` before,  but there doesn't seem to be much documentation and/or examples online,  and in my attempts I failed to apply an exit task for a few specific operations, as it complained that it has to be global for all.

**Comment by marrrcin** [comment]
This is related to https://github.com/kubeflow/pipelines/issues/6749

General workaround is to always return status code 0 from the pipeline steps and e.g. return some output (e.g. a string `OK` or `FAIL`) instead, which can be chained together with `dsl.Condition` to verify whether to continue the pipeline or not in a specific branch of the `ParallelFor`. 

**This workaround does not cover the overall pipeline status**, so no warning ⚠️ signs / red statuses - everything is green ✅

**Comment by marrrcin** [comment]
@yarnabrina , @chensun I've created a pull request implementing this behaviour - I would really appreciate your feedback on that [https://github.com/kubeflow/pipelines/pull/7373](https://github.com/kubeflow/pipelines/pull/7373).

**Comment by thesuperzapper** [comment]
I think most situations can be handled with an [`kfp.dsl.ExitHandler`](https://kubeflow-pipelines.readthedocs.io/en/stable/source/kfp.dsl.html#kfp.dsl.ExitHandler), which runs a single ContainerOp regardless of if the ContainerOps it wraps succeed or fail. 

But we might consider making functionality like ExitHandler more "implicit" by having an [airflow-style `trigger_rule`](https://airflow.apache.org/docs/apache-airflow/stable/concepts/dags.html#trigger-rules) flag on the ContainerOp. (Proposed in issue: https://github.com/kubeflow/pipelines/issues/2372)

**Comment by marrrcin** [comment]
@thesuperzapper it's hardly "most situations" in my opinion. Take an example of having branch-out scenario with ParallelFor as in my PR - it cannot be handled by ExitHandler, because there can be only one ExitHandler in the pipeline.

**Comment by marrrcin** [comment]
I don't fully understand where it's "most situations". Are multiple exit handlers now supported in KFP? As far as I can see, no https://github.com/kubeflow/pipelines/blob/e2687ce5c22455bbde0ababb3ad46588d5ed7939/sdk/python/kfp/compiler/compiler.py#L236 , so in common scenario of branch-out with ParallelFor (as in my PR) - it cannot be used.

**Comment by thesuperzapper** [comment]
@marrrcin I agree that only having a single ExitHandler is problematic, would allowing multiple also address your issue?

**Comment by marrrcin** [comment]
I don't think so - it's not a single job launched multiple times in parallel, it's a chain of consecutive jobs of which some might be allowed to fail - you can take a look at the screenshot in my PR (https://github.com/kubeflow/pipelines/pull/7373). Even having multiple exit handlers would not cover that imho.

**Comment by jszendre** [comment]
Exit handler ops cannot explicitly depend on any previous operations so they cannot be parameterized by outputs of previous operations or be guaranteed to run after previous steps.

My use case is having integration tests run that are themselves kubeflow pipelines and I would like to be able to verify that a task fails without the integration test failing. Configuring that in the dsl would be a lot cleaner than being included in application logic or directly in ci/cd.

**Comment by techwithshadab** [comment]
I also have a similar scenario, any work around this yet?

**Comment by mprahl** [comment]
The KFP maintainers discussed this and we agree that this should be prioritized.

The current thinking is if this was enabled on a task, we'd essentially mark the task as failed but continue running the rest of pipeline until completion. The pipeline would end up with a failed status though.

**Comment by rvolchek** [comment]
🏁 flagged for review, p0

**Comment by Sharkyii** [comment]
Dug into the code for the root cause: Argo's DAGTemplate.FailFast defaults to true, and KFP never overrides it anywhere in the compiler -that's why independent ParallelFor branches get cut short the moment one fails.

.ignore_upstream_failure() doesn't help here it's a per-task hook, not a FailFast fix.

@mprahl's design (mark the task failed, let the rest finish, end with a failed status) maps directly onto just flipping that flag.

Proposal: a pipeline-level opt-in (e.g. fail_fast=False) that sets FailFast: false on the compiled DAG when enabled, defaulting to current behavior. Small, contained, no changes needed to run-status aggregation.

@chensun @mprahl @HumairAK - does this match what you had in mind? If so, happy to go ahead and open a PR for it.
